src/test/java/org/opensearch/security/ssl/SslSettingsManagerReloadListenerTest.java

```diff
@@ -0,0 +1,256 @@
+                );
+                writeStore(keyStore, path(String.format("%s_keystore.p12", filePrefix)), keyStorePassword);
+            }
+        );
+    }
+
+    void reloadSslContextOnFilesChanged(final Settings settings, final CertificatesWriter certificatesWriter) throws Exception {
+        final var defaultHttpCertificates = generateCertificates();
+        final var defaultHttpKeyPair = defaultHttpCertificates.v1();
+        final var httpCaCertificate = defaultHttpCertificates.v2().v1();
+        final var httpAccessKeyAndCertificate = defaultHttpCertificates.v2().v2();
+
+        final var defaultTransportCertificates = generateCertificates();
+        final var defaultTransportKeyPair = defaultTransportCertificates.v1();
+        final var transportCaCertificate = defaultTransportCertificates.v2().v1();
+        final var transportAccessKeyAndCertificate = defaultTransportCertificates.v2().v2();
+
+        final var reloadHttpCertificates = randomBoolean();
+
+        certificatesWriter.write("http", httpCaCertificate, httpAccessKeyAndCertificate);
+        certificatesWriter.write("transport", transportCaCertificate, transportAccessKeyAndCertificate);
+
+        final var sslSettingsManager = new SslSettingsManager(TestEnvironment.newEnvironment(settings));
+        sslSettingsManager.addSslConfigurationsChangeListener(resourceWatcherService);
+
+        final var httpSslContextBefore = sslSettingsManager.sslContextHandler(CertType.HTTP).orElseThrow().sslContext();
+        final var transportSslContextBefore = sslSettingsManager.sslContextHandler(CertType.TRANSPORT).orElseThrow().sslContext();
+        final var transportClientSslContextBefore = sslSettingsManager.sslContextHandler(CertType.TRANSPORT_CLIENT)
+            .orElseThrow()
+            .sslContext();
+
+        final var filePrefix = reloadHttpCertificates ? "http" : "transport";
+        final var keyPair = reloadHttpCertificates ? defaultHttpKeyPair : defaultTransportKeyPair;
+        var caCertificate = reloadHttpCertificates ? httpCaCertificate : transportCaCertificate;
+        var keyAndCertificate = reloadHttpCertificates ? httpAccessKeyAndCertificate : transportAccessKeyAndCertificate;
+
+        if (randomBoolean()) {
+            caCertificate = certificatesRule.generateCaCertificate(
+                keyPair,
+                caCertificate.getNotBefore().toInstant(),
+                caCertificate.getNotAfter().toInstant().plus(365, ChronoUnit.DAYS)
+            );
+        } else {
+            keyAndCertificate = certificatesRule.generateAccessCertificate(
+                keyPair,
+                keyAndCertificate.v2().getNotBefore().toInstant(),
+                keyAndCertificate.v2().getNotAfter().toInstant().plus(365, ChronoUnit.DAYS)
+            );
+        }
+        certificatesWriter.write(filePrefix, caCertificate, keyAndCertificate);
```

Comment:
What happens if these files are not written simultaneously?

Reply:
The SSL context will be reloaded multiple times. To be honest, I don’t think there’s a way to implement this more efficiently.